nextcord/application_command.py

```diff
@@ -0,0 +1,883 @@
+        cog_parent: Optional[:class:`ClientCog`]
+            Cog parent, supplied as the first argument to the callback if provided.
+        name: :class:`str`
+            Name of the command on Discord. Defaults to the name of the callback.
+        description: :class:`str`
+            Description of the command on Discord, if applicable. Defaults to an empty string.
+        choices: Optional[Dict[:class:`str`, Any]]
+            A dictionary of choices. Keys are the Discord-facing name of the choice, the values are what Discord sends
+            back.
+        """
+        if not asyncio.iscoroutinefunction(callback):
+            raise TypeError("Callback must be a coroutine.")
+
+        self._parent: Union[ApplicationCommand, ApplicationSubcommand] = parent
+        self._callback: Callable = callback
+
+        self.type: Union[ApplicationCommandType, ApplicationCommandOptionType] = cmd_type
+        self.name: str = name
+        self.description: str = description
+        self.choices: Dict[str, Any] = choices
+
+        self.cog_parent: Optional[ClientCog] = cog_parent
+        self.arguments: Dict[str, CommandOption] = {}
+        self.children: Dict[str, ApplicationSubcommand] = {}
+        self._analyze_content()
+        self._analyze_callback()
+
+    def _analyze_content(self):
+        """This reads the content of itself and performs validation and changes to variables as needed."""
+        if self._parent and self._parent.type is ApplicationCommandOptionType.sub_command_group and self.children:
+            raise NotImplementedError("A subcommand can't have both subcommand parents and children! Discord does not"
+                                      "support this.")
+        if isinstance(self._parent, ApplicationCommand):
+            if self.children:
+                self.type = ApplicationCommandOptionType.sub_command_group
+            else:
+                self.type = ApplicationCommandOptionType.sub_command
+        if self.type is ApplicationCommandType.user or self.type is ApplicationCommandType.message:
+            self.description = ""
+        else:
+            if not self.description:
+                self.description = " "
+
+    def _analyze_callback(self):
+        """This reads the callback, performs validation, and changes variables as needed."""
+        if not self.name:
+            self.name = self._callback.__name__
+        first_arg = True
+
+        for value in signature(self.callback).parameters.values():
+            self_skip = value.name == "self"  # TODO: What kind of hardcoding is this, figure out a better way for self!
+            if first_arg:
+                # TODO: Is this even worth having?
+                if value.annotation is not value.empty and value.annotation is not Interaction:
+                    raise TypeError("The first argument in an Application Command should be typed as an Interaction")
```

Comment:
Likely this should be soft typed, no need to force typehints on end users imo. There loss

Reply:
+1, their fault if not typed right